libcore/StockDialogs.vala

```diff
@@ -26,50 +40,63 @@ namespace PF.Dialogs {
         /* Use default button type "CLOSE" */
         var dialog = new Granite.MessageDialog.with_image_from_icon_name (primary_text, secondary_text,
                                                                           "dialog-warning");
         return display_dialog (dialog, parent);
     }
 
-    /* Need to implement this until replace Gtk dialogs generated by marlin_file_operations do_run_simple_dialog () */
-    public void gtk_message_dialog_set_details_label (Gtk.MessageDialog dialog, string details_text) {
-        var details_view = new Gtk.TextView ();
-        details_view.border_width = 12;
-        details_view.buffer.text = details_text;
-        details_view.editable = false;
-        details_view.pixels_below_lines = 3;
-        details_view.wrap_mode = Gtk.WrapMode.WORD;
-        details_view.get_style_context ().add_class (Granite.STYLE_CLASS_TERMINAL);
-
-        var scroll_box = new Gtk.ScrolledWindow (null, null);
-        scroll_box.margin_top = 12;
-        scroll_box.min_content_height = 70;
-        scroll_box.add (details_view);
-
-        var expander = new Gtk.Expander.with_mnemonic (_("Show more _details"));
-        expander.resize_toplevel = true;
-        expander.add (scroll_box);
-        expander.show_all ();
-
-        var content_area = (Gtk.Box)(dialog.get_message_area ());
-        content_area.pack_start (expander, false, false, 0);
-
-        dialog.resizable = true;
-    }
-
     private Granite.MessageDialog display_dialog (Granite.MessageDialog dialog, Gtk.Window? parent) {
-        if (parent != null) {
+        if (parent != null && parent is Gtk.Window) {
             dialog.set_transient_for (parent);
         }
 
-        dialog.response.connect (() => {
+        dialog.response.connect_after (() => {
             dialog.destroy ();
         });
 
         dialog.show ();
         return dialog;
     }
+
+    public int run_simple_file_operation_dialog (Marlin.RunSimpleDialogData data) {
+        string image_name;
+        switch (data.message_type) {
+            case Gtk.MessageType.ERROR:
+                image_name = "dialog-error";
+                break;
+            case Gtk.MessageType.WARNING:
+                image_name = "dialog-warning";
+                break;
+            case Gtk.MessageType.QUESTION:
+                image_name = "dialog-question";
+                break;
+            default:
+                image_name = "dialog-information";
+                break;
+        }
+
+        var dialog = new Granite.MessageDialog.with_image_from_icon_name (data.primary_text,
+                                                                          data.secondary_text,
+                                                                          image_name,
+                                                                          Gtk.ButtonsType.NONE);
+
+        if (data.button_titles.length == 0) {
+            dialog.add_button (Gtk.Stock.CLOSE, 0);
```

Comment:
Gtk.Stock.CLOSE has been deprecated since 3.10. I think the solution here is probably to manually create a close button so we don't have to come back and change this for Gtk4

Reply:
Oops, the C code still uses GTK_STOCK_CANCEL a lot when calling the simple dialog.  I guess I should fix that as well.